File: 백준/Silver/2178. 미로 탐색/미로 탐색.cc
#include <bits/stdc++.h>	
using namespace std;

#define X first
#define Y second

int board[105][105];
int dist[105][105];
int n, m;

int dx[4] = { 1, 0, -1, 0 };
int dy[4] = { 0, 1, 0, -1 };

int main(void) {
	ios::sync_with_stdio(0);
	cin.tie(0);
	
	cin >> n >> m;

	fill(dist[0], dist[105], -1);

	string s;

	for (int i = 0; i < n; i++) {
		cin >> s;
		for (int j = 0; j < m; j++) {
			board[i][j] = s[j] - '0';
		}
	}

	queue<pair<int, int>> Q;
	Q.push({ 0, 0 });
	dist[0][0] = 0;

	while (!Q.empty()) {
		pair<int, int> cur = Q.front();
		Q.pop();
		for (int dir = 0; dir < 4; dir++) {
			int nx = cur.X + dx[dir];
			int ny = cur.Y + dy[dir];

			if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
			if (dist[nx][ny] != -1 || board[nx][ny] != 1) continue;

			dist[nx][ny] = dist[cur.X][cur.Y] + 1;
			Q.push({ nx, ny });
		}
	}

	cout << dist[n - 1][m - 1] + 1;
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.awt.*;
import java.util.*;
import java.io.*;

public class Main {

    static int N;
    static int M;
    static int[][] board;
    static int[][] dist;
    static int[] dx = new int[]{0, 1, 0, -1};
    static int[] dy = new int[]{1, 0, -1, 0};


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        board = new int[N][M];
        dist = new int[N][M];

        for(int i = 0; i < N; i++){
            Arrays.fill(dist[i], -1);
        }

        Queue<Point> q = new LinkedList<>();

        for(int i = 0; i < N; i++){
            String s = br.readLine();
            for(int j = 0; j < M; j++){
                board[i][j] = (int)(s.charAt(j) - '0');
            }
        }

        Point start = new Point(0, 0);
        q.add(start);
        dist[start.x][start.y] = 1;

        while(!q.isEmpty()){
            Point cur = q.poll();
            
            for(int dir = 0; dir < 4; dir++){
                int nx = cur.x + dx[dir];
                int ny = cur.y + dy[dir];

                if(nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
                if(dist[nx][ny] >= 0 || board[nx][ny] == 0) continue;

                dist[nx][ny] = dist[cur.x][cur.y] + 1;
                q.add(new Point(nx, ny));
            }
        }

        System.out.print(dist[N - 1][M - 1]);

    }
}
